### PR TITLE
Feature/eip 196

### PR DESCRIPTION
**Pull request by KonradStaniec**
Adding contracts for addition and scalar multiplication on bn128 curve.

Should be merged and reviewed after #490

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/crypto/zksnark/FieldElement.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    override def neg(a: Fp): Fp = Fp {
+      (-a.inner) mod P
+    }
+
+    override def isValid(a: Fp): Boolean = {
+      a.inner >= 0 && a.inner < P
+    }
+
+    override def isZero(a: Fp): Boolean =
+      a.inner.compareTo(BigInt(0)) == 0
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply src/main/scala/io/iohk/ethereum/crypto/zksnark/FieldElement.scala]: probably just to hasty during porting it :P will fix that

File: src/main/scala/io/iohk/ethereum/crypto/zksnark/BN128.scala [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  def isValidPoint(p1: Point[T]): Boolean =
+    p1.isValid && isOnCurve(p1)
+}
+
+object BN128Fp extends BN128[Fp] {
+  val Fp_B = Fp.B_Fp
+
+  override def createPointOnCurve(xx: ByteString, yy: ByteString)(implicit ev: FiniteField[Fp]): Option[Point[Fp]] = {
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply src/main/scala/io/iohk/ethereum/crypto/zksnark/BN128.scala]: In https://github.com/input-output-hk/mantis/pull/499/files#diff-357a7578811b3e0f2e07eaaed972773cR162 filter will return None if point is invalid

[the remark this answers is not in the document]
Comment (mmrozek) [reply src/main/scala/io/iohk/ethereum/crypto/zksnark/BN128.scala]: Sorry, I missed that.

**Review by agnikowa — APPROVED** [review]
Apart from weird formatting in `BN128FpSpec` (29, 37) and in `PrecompiledContracts` (245, 251-252, 255-256, 283) looks good to me

**Review by mmrozek — APPROVED** [review]

**Pull request merged**